packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -3655,18 +3655,103 @@ function commitPassiveUnmountOnFiber(finishedWork: Fiber): void {
       }
       break;
     }
-    // TODO: Disconnect passive effects when a tree is hidden, perhaps after
-    // a delay.
-    // case OffscreenComponent: {
-    //   ...
-    // }
+    case OffscreenComponent: {
+      const instance: OffscreenInstance = finishedWork.stateNode;
+      const nextState: OffscreenState | null = finishedWork.memoizedState;
+
+      const isHidden = nextState !== null;
+
+      if (
+        isHidden &&
+        instance.visibility & OffscreenPassiveEffectsConnected &&
+        // For backwards compatibility, don't unmount when a tree suspends. In
+        // the future we may change this to unmount after a delay.
+        (finishedWork.return === null ||
+          finishedWork.return.tag !== SuspenseComponent)
```

Comment:
Does this catch this case `<Suspense fallback={<Offscreen>...</Offscreen>}>{...}</Suspense>`?

Might be cleaner to just add another "mode" for now that user space can use too.

Reply:
It's for the internal Offscreen fiber that Suspense uses. We don't unmount passive effects when there's an avoided fallback.

Yeah I can add a "mode" for that. For the userspace API, I had imagined we might have a way to control the delay for when effects get detached, and one of the options could be "never".

Reply:
> Does this catch this case <Suspense fallback={<Offscreen>...</Offscreen>}>{...}</Suspense>?

Oh I see you're asking if there's a false positive. No because the fallback children are wrapped in an extra Fragment for reconciliation purposes.